Add custom k_diffusion scheduler in settings

KohakuBlueleaf opened the pull request:
Based on #10634, use settings instead of add new things in main interface.
Recommend to use with #10648 

![image](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/59680068/844ff3e0-60ab-4725-9852-95b97d740840)
![image](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/59680068/1144e636-2173-466d-81b1-41a9f6ad0ee9)

KohakuBlueleaf committed:
Add custom karras scheduler

KohakuBlueleaf committed:
remove debug print

KohakuBlueleaf committed:
Add dropdown for scheduler type

KohakuBlueleaf committed:
Change karras to kdiffusion

KohakuBlueleaf committed:
Replace karras by k_diffusion, fix gen info

KohakuBlueleaf committed:
only add metadata when k_sched is actually been used

KohakuBlueleaf committed:
remove not related code

KohakuBlueleaf committed:
Avoid loop import

KohakuBlueleaf committed:
Minor naming fixes

KohakuBlueleaf committed:
Add error information for recursion error

KohakuBlueleaf committed:
use sigma_max/min in model if sigma_max/min is 0

KohakuBlueleaf committed:
Better hint for user

Co-authored-by: catboxanon <122327233+catboxanon@users.noreply.github.com>

KohakuBlueleaf committed:
Add hint for custom k_diffusion scheduler

KohakuBlueleaf committed:
Merge branch 'custom-k-sched' of https://github.com/KohakuBlueleaf/stable-diffusion-webui into custom-k-sched

KohakuBlueleaf committed:
Use settings instead of main interface

KohakuBlueleaf commented:
@catboxanon Now I also have settings version

KohakuBlueleaf commented:
> You shouldn't need to modify the function signature of `StableDiffusionProcessing` and add those variables to the class -- take a look at how other parameters like CLIP skip are handled.

Oh I just take how s_tmax and s_churn works...
So these thing also should not work like that?

KohakuBlueleaf committed:
Use better way to impl

KohakuBlueleaf commented:
@catboxanon I use extra parameters now.
It looks more clean(?)

KohakuBlueleaf commented:
> Looks better -- be sure to update the xyz script by using `apply_override` instead too.

oh yes I forgot it

KohakuBlueleaf committed:
Fix xyz

KohakuBlueleaf committed:
Use type to determine if it is enable

KohakuBlueleaf committed:
Fix ruff error

KohakuBlueleaf commented:
@catboxanon done!

KohakuBlueleaf commented:
> `None` and/or `default` aren't really correct (a scheduler has to be used, and the default is only used for non-Karras samplers), which is why I think it should be renamed to `Automatic`, and is consistent with how the `sd_vae` and `cross_attention_optimization` settings are handled.

ok

KohakuBlueleaf committed:
Use automatic instead of None/default

AUTOMATIC1111 commented:
I have some changes that I have in mind for this:

- shorter names for infotext fields
- do not add infotext fields when their value is the default
- use actual `sigma_max` value in UI as default (it's 14.something)
- add fields to p.extra_generation_params in the very same place where they are added to sigmas_kwargs rather than in a different function

If you feel like it, you can add any of those into PR, otherwise I'll just merge this later when I have time to implement and test them.

KohakuBlueleaf commented:
> I have some changes that I have in mind for this:
> 
> * shorter names for infotext fields
> * do not add infotext fields when their value is the default (which would require the default values to be added in `parse_generation_parameters`)
> * use actual `sigma_max` value in UI as default (it's 14.something)
> * add fields to p.extra_generation_params in the very same place where they are added to sigmas_kwargs rather than in a different function
> 
> If you feel like it, you can add any of those into PR, otherwise I'll just merge this later when I have time to implement and test them.

I will do:
1. change "KDiffusion Scheduler xxx" to "KDiff Sched xxx", which will be shorter than other 3~5 longest names so I think it is acceptable.
2. combine 2 and 4, add extra parameter only if use non-default value.
3. use min=0.3 and max=14.6 as default value, which is the default value for SD models. But still take 0.0 as model default. (It is possible that some model will change their beta/alpha schedule, so the "default sigma" will be different from original SD)

KohakuBlueleaf committed:
improvements

See:
https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/10649#issuecomment-1561047723

KohakuBlueleaf committed:
use Schedule instead of Sched

KohakuBlueleaf commented:
@AUTOMATIC1111 I think for now it is better?
